packages/samples/headless-react/server/server.tsx

```diff
@@ -0,0 +1,64 @@
+import path from 'path';
+import fs from 'fs';
+
+import express from 'express';
+import ReactDOMServer from 'react-dom/server';
+
+import App from '../src/App';
+import {
+  buildSearchEngine,
+  getSampleSearchEngineConfiguration,
+  buildSearchStatus,
+  SearchEngine,
+} from '@coveo/headless';
+
+const PORT = 3000;
+const app = express();
+
+app.get('/', async (req, res) => {
+  const engine = buildSearchEngine({
+    configuration: getSampleSearchEngineConfiguration(),
+  });
+
+  renderServerSide(engine);
+  await firstSearchExecuted(engine);
+  const app = renderServerSide(engine);
+
+  const indexFile = path.resolve('./build/index.html');
+  fs.readFile(indexFile, 'utf8', (err, data) => {
+    if (err) {
+      console.error('Something went wrong:', err);
+      return res.status(500).send('Internal error');
+    }
+
+    const state = JSON.stringify(engine.state);
+    const page = data
+      .replace('<div id="root"></div>', `<div id="root">${app}</div>`)
+      .replace(
+        '<script id="ssr"></script>',
+        `<script id="ssr">window.HEADLESS_STATE = ${state}</script>`
```

Comment:
This may be [an XSS vulnerability](https://pragmaticwebsecurity.com/articles/spasecurity/json-stringify-xss.html).